guide/avoiding-unnecessary-configuration.md

```diff
@@ -0,0 +1,161 @@
+
+If you want to give use an output property of a task as the input to another task, you can just transform the `TaskProvider` into another `Provider` using `map` or `flatMap`. You can call `.configure` on a `TaskProvider` to lazily configure the task without realising it (avoid calling `.get()` on `TaskProvider`s).
+
+```java
+abstract class MyTask extends DefaultTask {
+    @OutputFile
+    abstract RegularFileProperty getOutputFile();
+}
+```
+
+```java
+TaskProvider<MyTask> myTask = project.getTasks().register("myTask", MyTask.class);
+
+otherTaskProvider.configure(task -> {
+    task.getInput().set(myTask.flatMap(MyTask::getOutputFile));
+});
+```
+
+This also has the benefit of wiring up the dependencies between the two tasks automatically (no need for `dependsOn`).
+
+## Accessing tasks lazily
+
+> Further reading: [Task Configuration Avoidance: Eager APIs to avoid](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#eager_apis_to_avoid)
+
+To access tasks lazily, you need to make sure to call lazy APIs that return `TaskProvider` rather than `Task`.
+
+### Getting single tasks
+
+#### Bad:
+
+These all return a `Task` so will realise the task and run all the lazy configuration on it immediately:
+
+```java
+project.getTasks().getByName("foo");
+project.getTasks().findByName("foo");
+project.getTasks().getByPath(":foo");
+project.getTasks().findByPath(":foo");
+project.getTasks().withType(Foo.class).getByName(":foo");
+```
+
+#### Good:
+
+The `named` method returns a `TaskProvider`, allowing lazy configuration.
+
+```java
+TaskProvider<?> task = project.getTasks().named("foo");
+
+TaskProvider<FooTask> task = project.getTasks()
+        .withType(Foo.class)
+        .named("foo");
+```
+
+### Iterating over tasks
+
+#### Bad:
+
+These will immediately realise newly registered tasks when they are added (even if added using `register`) and run all the lazy configuration on them:
+
+```java
+project.getTasks().whenTaskAdded(task -> ...);
+project.getTasks().whenObjectAdded(task -> ...);
+```
+
+These will realise _all_ tasks, even ones already registered:
+
+```java
+project.getTasks().all(task -> ...);
+project.getTasks().matching(task -> ...);
+project.getTasks().withType(Foo.class).matching(task -> ...);
+```
+
+#### Good:
+
+`configureEach` will lazily configure each task that already exists as well as new ones that get added:
+
+```java
+project.getTasks().configureEach(task -> ...)
+```
+
+It's a bit like running `.configure` on each `TaskProvider` without realising the tasks.
+
+To move from `matching`, you can just use an `if` statement inside the `configureEach` rather than prefiltering the tasks.
+
+## Lazy `Configuration` registration
+
+`Configuration` objects should also be lazily registered. This follows the same pattern as tasks above. For example, you `register` rather than `create` a `Configuration`:
```

Comment:
is this strictly true? I wasn't aware of any perf penalty from realising `Configuration`s?